### ISSUE TITLE
[wgsl] express layout for host-shareable matrix values

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#922]
This is forked from #773

**Comment by dneto0** [comment]
This is blocking conversion of CTS tests from GLSL to WGSL.

**Comment by dneto0** [comment]
How SPIR-V and Vulkan handles this:

As a design choice, the matrix type in SPIR-V is like the other pure value types: bool and numeric scalars and bool and numeric vectors.  There is only one instance of one of those types, when specified with the same parameterization.

For example, you can't have

         %float = OpTypeFloat 32
         %another_float_type = OpTypeFloat 32   ; same parameterization. error!

Similarly this extends to vectors:

         %float = OpTypeFloat 32
         %v4float = OpTypeVector %float 4
         %another = OpTypeVector %float 4   ;  same parameterization. error!

Matrices are the same way:

         %float = OpTypeFloat 32
         %v4float = OpTypeVector %float 4
         %m4x4 = OpTypeMatrix %v4float 4   ;  a 4x4 matrix
         %another4x4 = OpTypeMatrix %v4float 4  ; same parameterization. error!

However, when a matrix value appears in storage that is copied between host and device, you need to know its layout.  In particular, we assume that the layout of its component vector is fixed and unambiguous. Then the remaining information we need is:

- Is the matrix row major or column major?  
  - row major = the component vectors are the rows of the matrix. 
  - column major = the component vectors are the columns of the matrix.

- What is the relative byte offset from one component vector to the next.  This is the "matrix stride".  In a row major matrix, it's the offset from one row vector to the next row vector.  In a column major matrix, it's the offset from one column vector to the next column vector.

These layout attributes need to be recorded somewhere.   

For SPIR-V *arrays* that are in host-shareable memory, the _array stride_ is an attribute of the array type itself.  But that means you can have two arrays with same "parameterization" of element type and number-of-elements but with different strides, leading to two different types:

             OpDecorate %arr_a ArrayStride 4
             OpDecorate %arr_b ArrayStride 16
             %uint = OpTypeInt 32 0
             %uint_5 = OpConstant %uint 5
    
             ; Two array types, with same parameterization, but different layout attributes
             %arr_a = OpTypeArray %uint %uint_5  ; an array of 5 uints, stride 4
             %arr_b = OpTypeArray %uint %uint_5  ; another array of 5 uints, stride 16

For example, %arr_b might be usable in a uniform buffer object (UBO) and in a shader storage buffer object (SSBO).  But %arr_a might be usable in an SSBO but not a UBO, because traditionally UBOs require array strides to be a multiple of 16.

But, since we want the matrix type in SPIR-V to be unique only based on the component vector and number of those components, we have to express layout using a different strategy.  Instead of making layout part of the type, we push the layout information out to the thing that contains the matrix.

For Vulkan, each descriptor binding location must correspond to a Block-decorated struct type in the SPIR-V shader.  This is true for a simple buffer variable, and also applies to arrays of buffers.  (In the array of buffers case, in SPIR-V you declare an array of Block-decorated structs.)

So if the matrix is to have layout expressed for it, that matrix type must appear somewhere within the struct declaration.  If the matrix is directly a member of the structure, then we decorate that struct member with RowMajor or ColMajor, and with MatrixStride.

Here's an example, with the original GLSL:
```glsl
#version 450

layout(set=0, binding=0) buffer M { // storage buffer
  mat4x4 m;
} buf;

void main() {
  buf.m[0][1] = 1.0;
}
```
The relevant parts of the SPIR-V are:
```
               OpName %M "M"
               OpMemberName %M 0 "m"
               OpName %buf "buf"
               OpMemberDecorate %M 0 ColMajor   ; attach layout info to the member
               OpMemberDecorate %M 0 Offset 0   
               OpMemberDecorate %M 0 MatrixStride 16  
               OpDecorate %M Block
               OpDecorate %buf DescriptorSet 0
               OpDecorate %buf Binding 0
...

      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%mat4v4float = OpTypeMatrix %v4float 4
          %M = OpTypeStruct %mat4v4float   ; the block-decorated structure type
%_ptr_StorageBuffer_M = OpTypePointer StorageBuffer %M
        %buf = OpVariable %_ptr_StorageBuffer_M StorageBuffer

...

          %5 = OpLabel
         %18 = OpAccessChain %_ptr_StorageBuffer_float %buf %int_0 %int_0 %uint_1
               OpStore %18 %float_1
               OpReturn

```

The MatrixStride and ColMajor attributes are decorated on the structure member.  This helps during address calculation (at the AccessChain instruction defining the pointer value %18), because the base pointer %buf is a value of type pointer-to-structure-M.

That's the simple "member is a matrix" case.  What about aggregations of a matrix inside other things?  In the array case, SPIR-V chooses to propagate the matrix layout through the array-ness, all the way back to the structure member containing that array-of-matrices, or array-of-array-of-matrix, and so on.

Another example:

```glsl
#version 450

layout(set=0, binding=0) buffer M {
  mat2x2 arr[5][5];
  mat4x4 m;
} buf;

layout(location = 0) in mat4x4 min;

void main() {
  buf.arr[1][1][1][1] = 1.0;
  buf.m[0][0] = 1.0;
}
```
The relevant parts of the SPIR-V:
```
               OpName %M "M"
               OpMemberName %M 0 "arr"
               OpMemberName %M 1 "m"
               OpName %buf "buf"

 ; the inner array stride is 16 bytes. Each mat2x2 occupies 16 bytes.
 ; the inner array element size is 5 x 16 = 80.
               OpDecorate %_arr_mat2v2float_uint_5 ArrayStride 16

 ; the outer array's stride is 80 bytes; its element type is the inner array.
               OpDecorate %_arr__arr_mat2v2float_uint_5_uint_5 ArrayStride 80

               OpMemberDecorate %M 0 ColMajor
               OpMemberDecorate %M 0 Offset 0
               OpMemberDecorate %M 0 MatrixStride 8

; The matrix layout attributes ColMajor propagates through the array-ness
; out to the struct member.
               OpMemberDecorate %M 1 ColMajor          
               OpMemberDecorate %M 1 Offset 400
               OpMemberDecorate %M 1 MatrixStride 16
               OpDecorate %M Block
               OpDecorate %buf DescriptorSet 0
               OpDecorate %buf Binding 0
...

      %float = OpTypeFloat 32
    %v2float = OpTypeVector %float 2
%mat2v2float = OpTypeMatrix %v2float 2
       %uint = OpTypeInt 32 0
     %uint_5 = OpConstant %uint 5
%_arr_mat2v2float_uint_5 = OpTypeArray %mat2v2float %uint_5
%_arr__arr_mat2v2float_uint_5_uint_5 = OpTypeArray %_arr_mat2v2float_uint_5 %uint_5
    %v4float = OpTypeVector %float 4
%mat4v4float = OpTypeMatrix %v4float 4

; the block-decorated structure type
          %M = OpTypeStruct %_arr__arr_mat2v2float_uint_5_uint_5 %mat4v4float
%_ptr_StorageBuffer_M = OpTypePointer StorageBuffer %M
        %buf = OpVariable %_ptr_StorageBuffer_M StorageBuffer
...
          %5 = OpLabel
         %24 = OpAccessChain %_ptr_StorageBuffer_float %buf %int_0 %int_1 %int_1 %int_1 %uint_1
               OpStore %24 %float_1
         %26 = OpAccessChain %_ptr_StorageBuffer_float %buf %int_1 %int_0 %uint_0
               OpStore %26 %float_1
               OpReturn
```

The cases explained so far are sufficient to handle any matrix layout needs.
If a matrix is contained in structure that itself is a member inside a block-decorated struct, then the matrix layout is annotated on the member of the _inner_ structure type.
Similarly, if the matrix is in an array inside a sub-structure, then again we propagate its layout to the member of the _inner_ structure type.

The byte-oriented layouts apply to matrices in buffers.  Matrices passed as inputs and outputs along the pipeline occupy "location" slots, which uses its own set of rules for determining occupancy.   So ColMajor, RowMajor, and MatrixStride don't apply to matrices in input/output.

That's it for matrix physical layout in SPIR-V.

**Comment by dneto0** [comment]
Matrix physical layout was discussed in the [WGSL meeting of 2020-07-21](https://docs.google.com/document/d/17mqJR40z325U_jftDBxOYAjbzdFxKCQiIUXDEI8d9k4/edit#)

@litherum proposed that the mechanism for matrix layout should use the same principles as agreed for array layout.  #910  That is, attach an attribute to the array's _type_ (just like in SPIR-V).

I agreed to explore this option to see how it affects the design in WGSL, and how it affects translation.

Off the top of my head:
- Making layout information part of the matrix types means we now have multiple matrix types with same parameterization but differing only in layout
- I would propose that operating on matrix values only be allowed in a layout-free matrix type.  This entails converting at memory load/store boundaries between layout-rich matrix types in (host-shared) memory and layout-free types in non-host-shared storage (function, private, workgroup, input, output) and in "const" values.
- We should probably ban layout-rich matrix types in non-host-shared variables.
- Address calculation (computing a reference pointer) to a matrix in host-shared memory still operates in the layout-rich matrix type, until you reach into the component type.
  - Computing a pointer to the component vector of a layout-rich matrix type LM produces a pointer to the vector, and vectors don't have layout (because we assume they have fixed unambiguous layouts)

I'll have to think this through a little more.

**Comment by kvark** [comment]
This is all very complex... I'd love to have WGSL not making it more complicated than it has to be. So @litherum proposal to treat matrix column/row major flag in the same way as we handle stride (for matrices and arrays) resonates with me.

I guess one of the points I'm confused is *why* in SPIR-V the design choice was made to separate them.

So both stride and major-ness affect the layout of complex types inside host-shared buffers, right? From the point of view of the shader compiler, once something is read (or before it's written to), I assume the layout doesn't matter, and it's all placed in vector registers anyway. If that's true:
  - why do we need to have the array stride being a part of the type?
  - If I pass the array (by value) from one function to another, the stride shouldn't affect anything.
  - maybe both of these layout clarifies should just be decorations on the shareable struct fields, not types.

**Comment by dneto0** [comment]
No matter how you model it, the layout is a property of values when they are in storage.

A variable is a reference to storage; it's expresses that the implementation ensures the allocation of space, and the name denotes a reference (or pointer) to that space. The pointer has a type.  In WGSL (and SPIR-V, and LLVM), the pointer is then a handle that lets you do stuff: 

- load,
- store, 
- compute the address of an internal thing (if the pointed-at thing was composite)
   - and in turn, loads and stores to those sub-components

Each of those activities needs to know the layout, so the layout is a property of the pointer value somehow.    (Ok, if you squint then load and store can be lower level and only do a copy of bytes to/from registers. But type safety means the types must match).

You've got choices:
- annotate all that layout at the time you form the pointer.  That's awkward if the thing you're pointing at is a composite thing
- annotate the layout on the declaration of the type you're pointing to.   That's the choice for SPIR-V for arrays and structs.
- or don't annotate anything at all until you finally have a load or store.   
     - (I recall seeing an argument that LLVM should go toward this style. I don't know if that has been resolved.)
     - This way is pretty bad, in my opinion because now both programmer and implementation needs to track a bunch of facts across different parts of the program that all have to be consistent.  May as well make it part of the type.

A sane way to proceed is:
-  to describe layout on the types that you are pointing to
- the pointer value inherits the layout when pointing to that type
- once the value is loaded and "in registers" the layout no longer has meaning.  

So schematically, at declaration:

     var<buffer> x : Foo;  // Foo is a type with annotated layout.

But when you load it, transition to a layout-free type:

      const value : FooWithoutLayout = x;  //  FooWithoutLayout has same logical structure, but no layout annotations.

That violates our type-matching rules so far, but we can make up our own rules.  Like, perhaps invent a syntax:

    type FooWithoutLayout = [[no_layout]] Foo; 

or even

       const value : [[no_layout]]Foo = x;

**Comment by dneto0** [comment]
In a separate track of discussion, I explored some of the consequences of putting row_major and column_major onto the matrix type.

Let's talk about accessing the component vector.  In GLSL and WGSL, if M is a matrix then M[i] is a column vector.
Fun things happen when you access M[i] on a row-major matrix:
-  When you read it, it's a gather operation, i.e. you load the scalar components of the resulting column vector, and each component is MatrixStride bytes apart from the next one.
- When you write it, it's a scatter operation which reverses the process: write to scalars, MatrixStride bytes apart.

Here's a sample GLSL compute shader:

```glsl
#version 450

layout(set=0, binding=0) buffer A {
   layout(column_major) mat2x4 mcols;
   layout(row_major) mat2x4 mrows;
} inbound;

layout(set=0, binding=1) buffer B {
   vec4 first;
   vec4 second;
} outbound;

void main() {
  outbound.first = inbound.mcols[1];
  outbound.second = inbound.mrows[1];  // Read is a gather
  inbound.mcols[0] = inbound.mrows[0];
  inbound.mrows[1] = inbound.mcols[1]; // Write is a scatter
}
```

The first two assignments in `main` copy the second column vector from a column-major matrix to outbound.first, and the second column vector from a row-major matrix to outbound.second.

We should be able to break this down into steps where we 1. compute the pointers to read from, then 2. load them, then 3. write to the outputs.  
Something like:

```rust
   fn main() -> void {
       const read_first_ptr : ptr<buffer, vec4<f32>> = inbound.mcols[1];
       const read_second_ptr : ptr<buffer, vec4<f32>> = inbound.mrows[1];   // Whoops pointer looks the same

       const first_read : vec4<f32> = read_first_ptr;
       const second_read : vec4<f32> =read_second_ptr;  // Whoops this is a gather, but we can't tell so far.
       .... // then write it out
   }
```

The loading action on the two pointers are different, so I would argue that the types should be different.  I'm going to assert that if the permissible actions on two objects are different then their types should be different.

This argues to me that a pointer to a column _vector_ inside a laid-out matrix has a _stride_ (number of bytes between the vector's own components)
When the matrix is column-major, then a pointer to its component vector has a stride equal to the element scalar size.
When the matrix is row-major, then a pointer to its component vector has a stride equal to the MatrixStride. (!)

Rewriting the WGSL draft a bit, we'd have:
  
```rust
   struct A [[block]] {
      [[offset  0]] mcols : [[col_major]] mat2x4<f32>;
      [[offset 32]]  mrows : [[row_major]] mat2x4<f32>;
   };
   struct B [[block]] {
      [[offset 0]] first : vec4<f32>;
      [[offset 16]] second : vec4<f32>;
   };
   [[set 0, binding 0]] var<buffer> inbound : A;
   [[set 0, binding 1]] var<buffer> outbound : B;

   fn main() -> void {
       const read_first_ptr : [[stride 4]]ptr<buffer, vec4<f32>> = inbound.mcols[1];
       const read_second_ptr : [[stride 8]]ptr<buffer, vec4<f32>> = inbound.mrows[1];   // Pointers look different. Good

       const first_read : vec4<f32> = read_first_ptr;
       const second_read : vec4<f32> = read_second_ptr;  // Now we know this is a gather.  Ok.
       .... // then write it out
   }
```
It's *good* that we can see that the two pointers are different, because operating on them has different behaviour.

(Matrices can be really confusing.  So I wrote this example up as an Amber test script, which verifies the action of the shader.
See https://github.com/google/amber/pull/896/files)

**Comment by dneto0** [comment]
My thinking so far:
- It's sensible to annotate layout on the types that will be stored in host-shareable memory
- those layouts transfer to pointers we form to reference that storage.
- layouts cease to have meaning when the values are loaded into "registers" so we can operate on them. 
- when writing a value from "registers" out to host-shareable storage, you need to reacquire the layout information, so it's sensible to have that as part of the pointer type.
- It's sensible to have a shorthand for "this type is like that one but without any layout information", and to have special type matching rules for loads and stores that let you easily transition between in registers and in host-shareable storage.
- Accessing the component vector of a row-major matrix is fundamentally different from accessing a column major matrix; that should be reflected in the pointer type.   This implies a  "stride" on a pointer type, at least for the "pointing inside a matrix" case.

**Comment by dneto0** [comment]
MSL appears to not support row-major matrices. 

So another option is:
- only support column major-matrices
     - then accessing column vectors is a regular vector access.  Only row-major matrices have weird gather/scatter semantics.
- to keep things simple, matrix stride would equal the size/alignment of the column vector.  (vec3 is same size as vec4, like usual)  So you wouldn't have to mention it.

**Comment by dneto0** [comment]
I just checked, and it looks like a matrix in a UBO has the same layout constraints as in an SSBO.  So the "ban row-major matrices" option implies we can use a fixed layout for matrices and we wouldn't have to annotate any of RowMajor, ColMajor, or MatrixStride.  

The only challenge I can see is that matrix initializers in HLSL list the scalars in *row major* order.  See https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-matrix#examples

So that's a detail for HLSL front ends to sort out, and for us to be careful about. 

(I haven't yet consulted our internal folks who maintain the DXC SPIR-V code generator. )

**Comment by grorg** [comment]
Discussed at the [2020-07-28 meeting](https://docs.google.com/document/d/1Ai4Tgm0PKBln2YjDywmcWQ2QyhIyuwNzWLkbnQr_jdo/).

**Comment by dneto0** [comment]
At the meeting, agreed to:

- drop support for row-major matrices, at least for MVP
- only support column-major matrices, without padding *between* column vectors. (still have a length-3 vector have the size of length-4 vector)
- In this case, you don't need to express RowMajor, ColMajor, MatrixStride.

**Comment by dneto0** [comment]
Spec text landed for this as part of #1215 which described memory layouts.
